hs_core/views/resource_rest_api.py

```diff
@@ -0,0 +1,622 @@
+
+        add_language = True
+        for element in metadata:
+            if 'language' in element:
+                if 'code' in element['language']:
+                    add_language = False
+                else:
+                    metadata.remove(element)
+                break
+
+        if add_language:
+            metadata.append({'language': {'code': 'eng'}})
+
+        add_rights = True
+        for element in metadata:
+            if 'rights' in element:
+                if 'statement' in element['rights'] and 'url' in element['rights']:
+                    add_rights = False
+                else:
+                    metadata.remove(element)
+                break
+
+        if add_rights:
+            # add the default rights/license element
+            metadata.append({'rights':
+                                 {'statement': 'This resource is shared under the Creative Commons Attribution CC BY.',
+                                  'url': 'http://creativecommons.org/licenses/by/4.0/'
+                                 }
+                            })
+        try:
+            resource = hydroshare.create_resource(
+                    resource_type=resource_type,
+                    owner=self.request.user,
+                    title=res_title,
+                    edit_users=request.DATA.get('edit_users', None),
+                    view_users=request.DATA.get('view_users', None),
+                    edit_groups=request.DATA.get('edit_groups', None),
+                    view_groups=request.DATA.get('view_groups', None),
+                    keywords=None,
+                    metadata=metadata,
+                    files=request.FILES.values(),
+            )
+        except Exception as ex:
+            error_msg = {'Resource creation failed': ex.message}
+            raise ValidationError(detail=error_msg)
+
+        error_message = None
+        # no need of this if
+        if resource:
+            # Send post-create resource signal
+            post_create_resource.send(sender=resource_class, resource=resource, metadata=metadata, user=request.user,
+                                      validate_files=file_validation_dict)
+
+            # TODO: Use the utils.get_file_validation() here
+            if 'are_files_valid' in file_validation_dict:
+                if not file_validation_dict['are_files_valid']:
+                    error_message = file_validation_dict.get('message', None)
+                    if not error_message:
+                        error_message = "Uploaded file(s) failed validation."
+
+        else:
+            error_msg = 'Resource creation failed.'
+            raise ValidationError(detail=error_msg)
+
+        # prepare response data
+        rest_api_response.status = 'CREATED'
+        rest_api_response.status_code = status.HTTP_201_CREATED
+        if error_message:
+            rest_api_response.message = 'Resource created. %s.' % error_message
```

Comment:
In the front end we allow creation of resource even if the file validation fails (this is mostly related to metadata extraction from the uploaded file) and we may keep the file with the resource. Each resource type makes the decision what to do (keep or not keep) with the failed file.